pkg/vpcmodel/subnetsIPBlocks.go

```diff
@@ -0,0 +1,252 @@
+//      to get freeAddressesBlocks, we first copy splitByFiltersBlocks, then we remove the subnet already allocated addresses.
+//      (we get these addresses from the subnet reserved IP).
+//       when a free address is needed, we take the first address of the block and remove it from the free blocks list.
+
+// /////////////////////////////////////////////////////////////////////////////////////////
+// oneSubnetBlocks hold the blocks of one subnet
+type oneSubnetBlocks struct {
+	subnetOriginalBlock *ipblock.IPBlock // the block of the cidr of the subnet
+	// splitByFiltersBlocks are the atomic blocks induced by the filters, the union of this slice is the subnetOriginalBlock
+	splitByFiltersBlocks []*ipblock.IPBlock
+	// freeAddressesBlocks are the splitByFiltersBlocks minus the reserved IPs
+	// each block in splitByFiltersBlocks has a corresponding block at freeAddressesBlocks
+	freeAddressesBlocks []*ipblock.IPBlock
+	// fullyReservedBlocks is a bool per block.
+	// its true if all the addresses in the original block are reserved IP
+	// for these blocks there is no need to create private IPs
+	fullyReservedBlocks []bool
+}
+
+// SubnetsIPBlocks is a map from the subnet crn to the subnets block
+type SubnetsIPBlocks map[string]*oneSubnetBlocks
+
+// getSubnetsIPBlocks() is the main func that creates the subnetsBlocks, steps:
+// 1. get the subnets original blocks
+// 2. calculate the filters blocks
+// 3. calculate the splitByFiltersBlocks
+// 4. calculate the freeAddressesBlocks
+func GetSubnetsIPBlocks(rc *datamodel.ResourcesContainerModel,
+	filtersCidrs []map[string][]*string) (subnetsBlocks SubnetsIPBlocks, err error) {
+	subnetsBlocks = SubnetsIPBlocks{}
+	// gets the original blocks of the subnets:
+	if err := subnetsBlocks.getSubnetsOriginalBlocks(rc); err != nil {
+		return nil, err
+	}
+	// calc the filters blocks:
+	filtersBlocks, err := getFiltersBlocks(filtersCidrs)
+	if err != nil {
+		return nil, err
+	}
+	// calc the splitByFiltersBlocks:
+	subnetsBlocks.splitSubnetsOriginalBlocks(rc, filtersBlocks)
+	// calc the freeAddressesBlocks:
+	if err := subnetsBlocks.getSubnetsFreeBlocks(rc); err != nil {
+		return nil, err
+	}
+	return subnetsBlocks, nil
+}
+
+func (subnetsBlocks SubnetsIPBlocks) getSubnetsOriginalBlocks(rc *datamodel.ResourcesContainerModel) (err error) {
+	for _, subnetObj := range rc.SubnetList {
+		subnetsBlocks[*subnetObj.CRN] = &oneSubnetBlocks{}
+		subnetsBlocks[*subnetObj.CRN].subnetOriginalBlock, err = ipblock.FromCidr(*subnetObj.Ipv4CIDRBlock)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// splitSubnetsOriginalBlocks() splits the subnet's cidr(s) to (maximal) disjoint blocks -
+// such that each block is atomic w.r.t. the filters rules
+func (subnetsBlocks SubnetsIPBlocks) splitSubnetsOriginalBlocks(rc *datamodel.ResourcesContainerModel, filtersBlocks filtersBlocks) {
+	for _, subnetObj := range rc.SubnetList {
```

Comment:
why not just iterate the subnets in `subnetsBlocks` here?

Reply:
we use *subnetObj.VPC.CRN here